### ISSUE TITLE
ER_LOCK_DEADLOCK when multiple clients access the server

### ISSUE THREAD

**Issue by smathot** [open-cogsci/omm-server#123]
(@nclaidiere I'm moving this from open-cogsci/omm-client#16, because I'm fairly sure this is a server-side issue.) 

On a regular basis the experiment stops with the following error message:

```
item-stack: omm_loop[run].omm_sequence[run].announce[run].experiment[run].trial_sequence[run].logger[run]
time: Mon Apr 11 18:59:13 2022
exception type: FailedToSendJobResults
exception message: {"message":"updateparticipationssetstatus_id= 2, updated_at= '2022-04-11 16:58:19' wherestudy_id= 17 andparticipant_id= 29 - Deadlock found when trying to get lock; try restarting transaction","name":"Error","code":"ER_LOCK_DEADLOCK","status":500}
```

This does not depend on the experiment, the ID or the box (I have attached a file with a few of these messages on github). It happens mostly when it’s busy, is it a conflict when two trials from the same individual are started? We had one experiment that did not use the new file system and this didn’t show up.

Here are a few messages like this:

item-stack: omm_loop[run].omm_sequence[run].announce[run].experiment[run].trial_sequence[run].logger[run]
time: Mon Apr 11 18:59:13 2022
exception type: FailedToSendJobResults
exception message: {"message":"updateparticipationssetstatus_id= 2,updated_at= '2022-04-11 16:58:19' wherestudy_id= 17 andparticipant_id= 29 - Deadlock found when trying to get lock; try restarting transaction","name":"Error","code":"ER_LOCK_DEADLOCK","status":500}

item-stack: omm_loop[run].omm_sequence[run].announce[run].experiment[run].trial_sequence[run].logger[run]
time: Tue Apr 12 08:39:38 2022
exception type: FailedToSendJobResults
exception message: {"message":"updateparticipationssetstatus_id= 2,updated_at= '2022-04-12 06:38:43' wherestudy_id= 17 andparticipant_id= 39 - Deadlock found when trying to get lock; try restarting transaction","name":"Error","code":"ER_LOCK_DEADLOCK","status":500}

item-stack: omm_loop[run].omm_sequence[run].announce[run].experiment[prepare].request_job[prepare]
time: Tue Apr 12 09:14:12 2022
exception type: FailedToSetJobStates
exception message: {"message":"updateparticipationssetstatus_id= 2,updated_at= '2022-04-12 07:13:16' wherestudy_id= 23 andparticipant_id= 34 - Deadlock found when trying to get lock; try restarting transaction","name":"Error","code":"ER_LOCK_DEADLOCK","status":500}

item-stack: omm_loop[run].omm_sequence[run].announce[run].experiment[run].trial_sequence[run].logger[run]
time: Tue Apr 12 09:15:55 2022
exception type: FailedToSendJobResults
exception message: {"message":"updateparticipationssetstatus_id= 2,updated_at= '2022-04-12 07:15:00' wherestudy_id= 17 andparticipant_id= 30 - Deadlock found when trying to get lock; try restarting transaction","name":"Error","code":"ER_LOCK_DEADLOCK","status":500}

openmonkeymind._exceptions.FailedToSendJobResults: {"message":"update participations set status_id = 2, updated_at = '2022-04-12 11:34:47' where study_id = 18 and participant_id = 41 - Deadlock found when trying to get lock; try restarting transaction","name":"Error","code":"ER_LOCK_DEADLOCK","status":500}

item-stack: omm_loop[run].omm_sequence[run].announce[run].experiment[run].trial_sequence[run].logger[run]
time: Tue Apr 12 13:29:46 2022
exception type: FailedToSendJobResults
exception message: {"message":"updateparticipationssetstatus_id= 2,updated_at= '2022-04-12 11:28:50' wherestudy_id= 18 andparticipant_id= 41 - Deadlock found when trying to get lock; try restarting transaction","name":"Error","code":"ER_LOCK_DEADLOCK","status":500}

item-stack: omm_loop[run].omm_sequence[run].announce[run].experiment[run].trial_sequence[run].logger[run]
time: Tue Apr 12 13:26:09 2022
exception type: FailedToSendJobResults
exception message: {"message":"updateparticipationssetstatus_id= 2,updated_at= '2022-04-12 11:25:13' wherestudy_id= 18 andparticipant_id= 38 - Deadlock found when trying to get lock; try restarting transaction","name":"Error","code":"ER_LOCK_DEADLOCK","status":500}

item-stack: omm_loop[run].omm_sequence[run].announce[run].experiment[prepare].request_job[prepare]
time: Tue Apr 12 13:45:35 2022
exception type: FailedToSetJobStates
exception message: {"message":"updateparticipationssetstatus_id= 2,updated_at= '2022-04-12 11:44:39' wherestudy_id= 18 andparticipant_id= 38 - Deadlock found when trying to get lock; try restarting transaction","name":"Error","code":"ER_LOCK_DEADLOCK","status":500}

item-stack: omm_loop[run].omm_sequence[run].announce[run].experiment[run].trial_sequence[run].logger[run]
time: Tue Apr 12 13:46:36 2022
exception type: FailedToSendJobResults
exception message: {"message":"updateparticipationssetstatus_id= 2,updated_at= '2022-04-12 11:45:39' wherestudy_id= 17 andparticipant_id= 40 - Deadlock found when trying to get lock; try restarting transaction","name":"Error","code":"ER_LOCK_DEADLOCK","status":500}

**Comment by dschreij** [comment]
Thanks. Will look into this.

**Comment by dschreij** [comment]
What endpoint are you calling here. Is it `PATCH /participants/:identifier/:job_id/result`?

**Comment by dschreij** [comment]
I think I found the cause for the deadlock. It is caused by the section https://github.com/open-cogsci/omm-server/blob/develop/app/Models/Study.js#L362-L374
After every trial, the status of the study is checked and either set to 'in progress' or finished. The problem is that the status is even changed to 'in progress' when the study already has that status, which is an unnecessary write operation of course, and the write operations lock the record. My guess is that when 2 participants complete a trial for the same study, the deadlock can occur (but that's a very in short time window all things considered... wow). I think the fix is simple. I've added a condition to the query that only records should be updated that don't already have the current status:

```js
await this.participants()
        .pivotQuery()
        .where('participant_id', ptcpID)
        .whereNot('status_id', statusID) // <-- new line
        .update({ status_id: statusID })
```
This should prevent the study record to be locked by two simultaneous queries, as the record no longer gets locked for updating if it already has the desired status. I do have difficulties reproducing the deadlock locally (even with an automated test), so I have no idea if this will really fix the issue. Would it be easy to check for you if it still occurs?

**Comment by nclaidiere** [comment]
Hi Daniel, great! we can test it soon, probably next week. Do I just need to update the server?

**Comment by dschreij** [comment]
Yes! I also did a quick release so you can also just use the docker image by updating it with docker-compose pull

**Comment by smathot** [comment]
Excellent, thanks!

**Comment by nclaidiere** [comment]
Hi, the deadlock issue is back. I changed the experiment for something much quicker, choosing between 2 stims, and got the error back. But I don't think I have the last update.

**Comment by dschreij** [comment]
Yes, I think you need version 0.7.4 where this should be fixed

**Comment by nclaidiere** [comment]
Ok great, I'll let you know how this goes.

Le jeu. 19 mai 2022, 08:43, Daniel Schreij ***@***.***> a
écrit :
